packages/core/src/routes/interaction/middleware/koa-interaction-hooks.ts

```diff
@@ -1,31 +1,79 @@
-import { trySafe } from '@silverhand/essentials';
+import { type InteractionEvent } from '@logto/schemas';
+import { conditional, trySafe } from '@silverhand/essentials';
 import type { MiddlewareType } from 'koa';
 import type { IRouterParamContext } from 'koa-router';
 
-import type TenantContext from '#src/tenants/TenantContext.js';
+import type Libraries from '#src/tenants/Libraries.js';
 
 import { getInteractionStorage } from '../utils/interaction.js';
 
 import type { WithInteractionDetailsContext } from './koa-interaction-details.js';
 
+export type InteractionHookContext = {
+  event: InteractionEvent;
+  sessionId?: string;
+  applicationId?: string;
+};
+
+export type InteractionHookResults = {
+  userId: string;
+};
+
+type AssignInteractionHookDetails = (details: InteractionHookResults) => void;
+
+export type WithInteractionHooksContext<
+  ContextT extends IRouterParamContext = IRouterParamContext
+> = ContextT & { assignInteractionHookResults: AssignInteractionHookDetails };
+
+/**
+ * The factory to create a new interaction hook middleware function for the submit interaction route.
+ * Submit interaction related event hooks will be triggered once we got the interaction hook results.
+ * Use `assignInteractionHookResults` to assign the interaction hook results.
+ */
 export default function koaInteractionHooks<
   StateT,
-  ContextT extends WithInteractionDetailsContext<IRouterParamContext>,
+  ContextT extends WithInteractionDetailsContext,
   ResponseT
 >({
-  provider,
-  libraries: {
-    hooks: { triggerInteractionHooksIfNeeded },
-  },
-}: TenantContext): MiddlewareType<StateT, ContextT, ResponseT> {
+  hooks: { triggerInteractionHooks },
+}: Libraries): MiddlewareType<StateT, WithInteractionHooksContext<ContextT>, ResponseT> {
   return async (ctx, next) => {
     const { event } = getInteractionStorage(ctx.interactionDetails.result);
+    const {
+      interactionDetails: {
+        jti: sessionId,
+        params: { client_id },
+      },
+      header: { 'user-agent': userAgent },
+    } = ctx;
+
+    // Predefined interaction hook context
+    const interactionHookContext: InteractionHookContext = {
+      event,
+      sessionId,
+      applicationId: conditional(typeof client_id === 'string' && client_id),
+    };
+
+    // eslint-disable-next-line @silverhand/fp/no-let, no-undef-init, unicorn/no-useless-undefined
+    let interactionHookResults: InteractionHookResults | undefined = undefined;
```

Comment:
optional ctx.intercationHookResult ?

Reply:
also, these hook context fields are not needed if interactionHookResults is empty?

Reply:
> optional ctx.intercationHookResult ?
Update the `interactionHookResults` with the type `Optional<InteractionHookResults`. It may not be a good idea to put this value into the ctx since it may be modified unexpectedly, and we want the user can only setup it up by `assignInteractionHookResults`

Reply:
> also, these hook context fields are not needed if interactionHookResults is empty?

We don't know if the context is needed. and we store them first since the details may lose some data when processing the interaction in the later logic.

Reply:
@simeng-li